### PR TITLE
Durable Nonce - Authorized Noncer

### LINKED ISSUE #7305: Add Nonce::Authorize instruction/support
There's a desire by partners to support single-signature Transactions that utilize a Nonce account.   Currently the Nonce account itself must be a signer, which prohibits such a Transaction.   This would be possible if we had a Nonce::Authorize instruction that behaved like the Stake::Authorize instruction, to authorize a system account to act as the authorized *noncer*.

### PR DESCRIPTION
**Pull request by t-nelson**
#### Problem

There is no mechanism for an account to delegate noncing authority within the durable nonce program

#### Summary of Changes

Add an "Authorized Noncer"
 - Specify authority within the `Initialize` instruction
 - Change authority with an `Authorize` instruction

Fixes #7305

### CODE REVIEW

File: sdk/src/nonce_state.rs [hunk 1]
```diff
@@ -12,12 +12,54 @@ use crate::{
 use serde_derive::{Deserialize, Serialize};
 use std::collections::HashSet;
 
+#[derive(Debug, Serialize, Deserialize, PartialEq, Clone, Copy)]
+pub enum AuthorityRole {
```
Comment (rob-solana) [inline sdk/src/nonce_state.rs]: this is overkill

**Comment by t-nelson** [comment]
10-4!

**Comment by t-nelson** [comment]
De-future-proofed and fixed up CLI

File: cli/src/cli.rs [hunk 2]
```diff
@@ -109,6 +109,7 @@ pub enum CliCommand {
     CreateNonceAccount {
         nonce_account: KeypairEq,
         lamports: u64,
+        authorized_noncer: Option<Pubkey>,
```
Comment (mvines) [inline cli/src/cli.rs]: `authorized_noncer` is kinda comical, perhaps `authorized_user` would be more approachable to a future reader that wasn't in the room when the Noncer joke was first cracked?

File: cli/src/nonce.rs [hunk 3]
```diff
@@ -207,6 +217,7 @@ pub fn process_create_nonce_account(
     config: &CliConfig,
     nonce_account: &Keypair,
     lamports: u64,
+    authorized_noncer: &Option<Pubkey>,
```
Comment (mvines) [inline cli/src/nonce.rs]: Can we do `Option<&Pubkey>`?    It aways looks so weird when you see `&None`.

File: cli/src/cli.rs (hunk 2 above)
Comment (t-nelson) [reply cli/src/cli.rs]: The humor was what I used to justify the long name :laughing:

Comment (mvines) [reply cli/src/cli.rs]: I personally like it.   I sadly feel it'll get lost in translation though

File: cli/tests/nonce.rs [hunk 4]
```diff
@@ -66,6 +66,7 @@ fn test_nonce() {
     config_payer.command = CliCommand::CreateNonceAccount {
         nonce_account: read_keypair_file(&keypair_file).unwrap().into(),
         lamports: 1000,
+        authorized_noncer: None,
```
Comment (jackcmay) [inline cli/tests/nonce.rs]: How about a test to verify that `authorized_noncer: Some(`

File: cli/src/cli.rs (hunk 2 above)
Comment (jackcmay) [reply cli/src/cli.rs]: Now I need to know, and for posterity, what was the joke?

**Comment by jackcmay** [comment]
No CLI support for `authorize`?

File: cli/src/cli.rs (hunk 2 above)
Comment (mvines) [reply cli/src/cli.rs]: You didn't miss much.  A Stake account has a Staker so a Nonce account should have a Noncer.   
😆😆😆😆 Get it?  😆😆😆😆

File: cli/tests/nonce.rs (hunk 4 above)
Comment (t-nelson) [reply cli/tests/nonce.rs]: Ah, yeah!  I guess I just did one for the parser.  Will add

File: cli/src/cli.rs (hunk 2 above)
Comment (jackcmay) [reply cli/src/cli.rs]: 👎

File: sdk/src/nonce_instruction.rs [hunk 5]
```diff
@@ -119,6 +132,14 @@ pub fn withdraw(nonce_pubkey: &Pubkey, to_pubkey: &Pubkey, lamports: u64) -> Ins
     )
 }
 
+pub fn authorize(nonce_pubkey: &Pubkey, authorized_noncer: &Pubkey) -> Instruction {
+    Instruction::new(
+        id(),
+        &NonceInstruction::Authorize(*authorized_noncer),
+        vec![AccountMeta::new(*nonce_pubkey, true)],
```
Comment (t-nelson) [inline sdk/src/nonce_instruction.rs]: @rob-solana the account key does not in fact need to sign (and may be unable to in the case of a derived address) here, correct?  Same for `Nonce` instruction above?

File: cli/src/cli.rs (hunk 2 above)
Comment (CriesofCarrots) [reply cli/src/cli.rs]: Is it a reference to the urban-dictionary definition of nonce? Or...?

File: sdk/src/nonce_state.rs [hunk 6]
```diff
@@ -141,13 +151,32 @@ impl<'a> NonceAccount for KeyedAccount<'a> {
...
     }
+
+    fn authorize(
+        &mut self,
+        authorized_noncer: &Pubkey,
+        signers: &HashSet<Pubkey>,
+    ) -> Result<(), InstructionError> {
+        match self.state()? {
+            NonceState::Initialized(meta, nonce) => {
+                if !signers.contains(&meta.authorized_noncer) {
```
Comment (t-nelson) [inline sdk/src/nonce_state.rs]: @rob-solana should the account owner also have authority to authorize?  Would enable revocation in the non-derived address case

**Comment by t-nelson** [comment]
> No CLI support for `authorize`?

Separate step https://github.com/solana-labs/solana/issues/7419

File: cli/src/cli.rs (hunk 2 above)
Comment (mvines) [reply cli/src/cli.rs]: OMG no!

Comment (CriesofCarrots) [reply cli/src/cli.rs]: Oh, good. I don't get it, then, but 🤷‍♀

Comment (t-nelson) [reply cli/src/cli.rs]: Dear god...

Nonce = Number used ONCE !

File: sdk/src/nonce_instruction.rs (hunk 5 above)
Comment (t-nelson) [reply sdk/src/nonce_instruction.rs]: Actually looks like more plumbing will be needed for this.  There doesn't seem to be a way to add a signer that isn't specified in an `AccountMeta` of one of the instructions

File: sdk/src/nonce_state.rs (hunk 6 above)
Comment (rob-solana) [reply sdk/src/nonce_state.rs]: create+init(Noncer)
there-after only the Noncer can/needs to sign

File: cli/src/nonce.rs [hunk 7]
```diff
@@ -231,7 +242,18 @@ pub fn process_create_nonce_account(
         .into());
     }
 
-    let ixs = create_nonce_account(&config.keypair.pubkey(), &nonce_account_pubkey, lamports);
+    let authorized_noncer = if let Some(authorized_noncer) = authorized_noncer {
```
Comment (rob-solana) [inline cli/src/nonce.rs]: unwrap_or(nonce_account_pubkey) ?

File: sdk/src/nonce_instruction.rs (hunk 5 above)
Comment (rob-solana) [reply sdk/src/nonce_instruction.rs]: go copy the code that stake and vote use for building the account metas

File: cli/src/cli.rs (hunk 2 above)
Comment (garious) [reply cli/src/cli.rs]: Please take this to Discord's terminology channel before this feature (not PR) ships. Names like this suggest the implementation may only be good enough for only whomever is asking for some consequence of it.  The name should stand on its own. Once pinned down, the terms should go into the book, preferably BEFORE the implementation.

**Comment by t-nelson** [comment]
I think all of the boxed are ticked here now

File: cli/src/nonce.rs [hunk 8]
```diff
@@ -22,6 +22,15 @@ pub trait NonceSubCommands {
     fn nonce_subcommands(self) -> Self;
 }
 
+fn nonce_authority_arg<'a, 'b>() -> Arg<'a, 'b> {
+    Arg::with_name("nonce_authority")
+        .long("authorized-noncer")
```
Comment (mvines) [inline cli/src/nonce.rs]:
```suggestion
        .long("nonce-authority")
```

File: cli/src/nonce.rs [hunk 9]
```diff
@@ -52,6 +61,14 @@ impl NonceSubCommands for App<'_, '_> {
                         .takes_value(true)
                         .possible_values(&["SOL", "lamports"])
                         .help("Specify unit to use for request"),
+                )
+                .arg(
+                    Arg::with_name("nonce_authority")
+                        .long("authorized-noncer")
```
Comment (mvines) [inline cli/src/nonce.rs]:
```suggestion
                        .long("nonce-authority")
```

File: cli/src/nonce.rs [hunk 10]
```diff
@@ -413,7 +454,33 @@ mod tests {
...
                     nonce_account: read_keypair_file(&keypair_file).unwrap().into(),
-                    lamports: 50
+                    nonce_authority: nonce_account_pubkey,
+                    lamports: 50,
+                },
+                require_keypair: true
+            }
+        );
+
+        // Test CreateNonceAccount SubCommand with authorized noncer
```
Comment (mvines) [inline cli/src/nonce.rs]:
```suggestion
        // Test CreateNonceAccount SubCommand with nonce authority
```

File: cli/src/nonce.rs [hunk 11]
```diff
@@ -496,12 +588,37 @@ mod tests {
...
+
+        // Test WithdrawFromNonceAccount Subcommand with authority
+        let test_withdraw_from_nonce_account = test_commands.clone().get_matches_from(vec![
+            "test",
+            "withdraw-from-nonce-account",
+            &keypair_file,
+            &nonce_account_string,
+            "42",
+            "lamports",
+            "--authorized-noncer",
```
Comment (mvines) [inline cli/src/nonce.rs]:
```suggestion
            "--nonce-authority",
```

File: cli/tests/nonce.rs [hunk 12]
```diff
@@ -53,6 +53,63 @@ fn test_nonce() {
...
+        &mut config_nonce,
+        &nonce_keypair_file,
+        &authority_keypair_file,
+    );
+
+    server.close().unwrap();
+    remove_dir_all(ledger_path).unwrap();
+}
+
+fn full_battery_tests(
```
Comment (mvines) [inline cli/tests/nonce.rs]: full battery?

File: sdk/src/nonce_instruction.rs [hunk 13]
```diff
@@ -63,12 +63,25 @@ pub enum NonceInstruction {
     ///     0 - A NonceAccount in the Uninitialized state
     ///     1 - RecentBlockHashes sysvar
     ///     2 - Rent sysvar
-    Initialize,
+    ///
+    /// The `Pubkey` parameter specifies the entity authorized to execute nonce
+    /// instruction on the account
+    Initialize(Pubkey),
```
Comment (mvines) [inline sdk/src/nonce_instruction.rs]: Passing Pubkeys in as instruction data is clunky, it requires custom serde code.  Instead just add the pubkey as one of the required non-signer keys

File: cli/src/nonce.rs (hunk 8 above)
Comment (t-nelson) [reply cli/src/nonce.rs]: Oof.  `sed` fail.

File: sdk/src/nonce_instruction.rs [hunk 14]
```diff
@@ -63,12 +63,25 @@ pub enum NonceInstruction {
...
+
+    /// `Authorize` changes the entity authorized to execute nonce instructions
+    /// on the account
+    ///
```
Comment (mvines) [inline sdk/src/nonce_instruction.rs]: Can you indicate who needs to sign the transaction here.  I assume it's the current authorized pubkey

File: cli/tests/nonce.rs (hunk 12 above)
Comment (t-nelson) [reply cli/tests/nonce.rs]: "any large group or series of related things"

File: sdk/src/nonce_instruction.rs (hunk 13 above)
Comment (jackcmay) [reply sdk/src/nonce_instruction.rs]: Saves having to pull that account, clone its data, etc...

Comment (t-nelson) [reply sdk/src/nonce_instruction.rs]: Left this one for the time being.  Seems like the runtime savings should take precedent?  wdyt @mvines ?

Comment (mvines) [reply sdk/src/nonce_instruction.rs]: Ok fine.  You win this round. 🤼‍♂

**Pull request merged**